Add Warp Payload Types (#2116)
ava-labs/avalanchego: vms/platformvm/warp/payload/README.md

# Payload

> An Avalanche Unsigned Warp Message already includes a `networkID`, `sourceChainID`, and `payload` field. The `payload` field is parsed into one of the types included in this package to be handled by the EVM.

Reviewer: ```suggestion
An Avalanche Unsigned Warp Message already includes a `networkID`, `sourceChainID`, and `payload` field. The `payload` field is parsed into one of the types included in this package to be further handled by the VM.
```

## AddressedPayload

AddressedPayload:
```
+---------------------+----------+-------------------+
|             codecID :   uint16 |           2 bytes |
+---------------------+----------+-------------------+
|              typeID :   uint32 |           4 bytes |
+---------------------+----------+-------------------+
|       sourceAddress : [20]byte |          20 bytes |
+---------------------+----------+-------------------+
|             payload :   []byte |  4 + len(payload) |
+---------------------+----------+-------------------+
                                 | 30  + len(payload) |
                                 +-------------------+
```

- `codecID` is the codec version used to serialize the payload and is hardcoded to `0x0000`
- `typeID` is the payload type identifier and is `0x00000000` for `AddressedPayload`

> - `sourceAddress` is the address that called `sendWarpPrecompile` on the source chain

Reviewer: I think we should make this more generic. Seems a bit arbitrary to force 20 byte addresses for all VMs. (or assume precompile language here).
(Actually seems the code is updated, should we update this document/diagram to match?)

> - `sourceAddress` is the address that called `sendWarpPrecompile` on the source chain

Reviewer 2: ```suggestion
- `sourceAddress` is the address that sent this message from the source chain
```

- `payload` is an arbitrary byte array payload

> ## BlockHash

Reviewer 3: Should we call this `ProofHash`?

Reviewer 3: As we discussed, this may not be strictly a block hash

Reviewer 4: This is currently used as a `BlockHash` and making it more generic means that the caller would need to know what kind of hash it's referring to (some context is already needed ofc), so I opted to leave this as `BlockHash` for now.

If we want to change the name in the future, it is backwards compatible to do so, so I'd prefer to leave this until we actually use it as something other than a `BlockHash`.

Reviewer 3: As you alluded to, blocks aren't "universal" so the same problem applies to them.

I think locking this terminology in when we communicate it broadly will make us appear a little short-sighted, whereas using the `Proof`/`ProofRoot` moniker is forward-looking.

Reviewer 4: Tbh I'd expect using a `BlockHash` and using that to prove out the rest of the header contents to be simpler.

Do you see this as a more targeted? If a VM would support multiple different types of proofs, how do you foresee them providing context as to what it's proving? Do you foresee them being registered under different typeIDs?

Reviewer 3: I guess I was really excited by the generality/simplicity of using an arbitrary proof structure. For example, it may be way easier to write a solidity contract to verify a trie of warp message IDs rather than a HyperSDK block. That approach also provides a generic interface that any VM can implement and use previously tested/implemented contracts (which I think will be a huge unlock).

We can enshrine "type 0" of the type to be "block hash" if you want to call that out.

Reviewer 3: In terms of using "hash" in the name, I'm not sure what it should be called. I'm not sure all "proof roots" are "hashes" (maybe they are)?

Reviewer 4: Ok I guess the context will be up to the Solidity contracts to know what they are talking to based on the chainID, so that they know how it should be interpreted.

As for naming, I'm fine with `Hash` or `Proof`

> BlockHashPayload:

Reviewer 2: Is there a reason this is `BlockHashPayload` rather than a `Hash` payload? Seems unnecessarily specific.

Reviewer 4: This is meant to be specific to the BlockHash as opposed to an arbitrary hash. If we changed to an arbitrary hash, the caller would need to have a different way to differentiate between hash types 🤷 .

I don't think there's a need for an arbitrary hash payload at this time, so lmk if you'd prefer I change it anyways.

Reviewer 4: Going to leave this as `BlockHash` after removing the `Payload` suffix.
